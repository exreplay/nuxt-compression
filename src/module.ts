import { readFileSync, writeFileSync } from 'fs'
import { defineNuxtModule } from '@nuxt/kit'
import viteCompression from 'vite-plugin-compression'
import { globbySync } from 'globby'
import { virtual } from './virtual'

export type ViteCompressionOptions = Parameters<typeof viteCompression>[0];

export interface ModuleOptions {
  viteCompression?: ViteCompressionOptions;
}

export default defineNuxtModule<ModuleOptions>({
  meta: {
    name: 'compression',
    configKey: 'compression'
  },
  defaults: {
    viteCompression: {
      algorithm: 'brotliCompress',
      filter: /\.(js|mjs|json|css|html)$/i
    }
  },
  setup (options, nuxt) {
    const filter = options.viteCompression.filter
    let fileExtension = ''
    let contentEncoding = ''

    switch (options.viteCompression.algorithm) {
      case 'gzip':
        fileExtension = 'gz'
        contentEncoding = 'gzip'
        break
      case 'brotliCompress':
        fileExtension = 'br'
        contentEncoding = 'br'
        break
    }

    nuxt.hook('vite:extend', (context) => {
      context.config.plugins?.push(
        viteCompression(options.viteCompression)
      )
    })

    nuxt.hook('nitro:build:before', (nitro) => {
      nitro.options.rollupConfig?.plugins?.push(
        virtual(
          {
            '#internal/nitro/virtual/public-assets-node': `
import { promises as fsp, constants } from 'fs'
import { resolve } from 'pathe'
import { dirname } from 'pathe'
import { fileURLToPath } from 'url'
import assets from '#internal/nitro/virtual/public-assets-data'

${typeof filter === 'function' ? `const filterFunction = ${filter.toString()}` : ''}

export async function readAsset (id, res) {
  const serverDir = dirname(fileURLToPath(import.meta.url))
  let assetPath = resolve(serverDir, assets[id].path)

  if (${typeof filter === 'function' ? 'filterFunction(assetPath)' : '/\\.(js|mjs|json|css|html)$/i.test(assetPath)'}) {
    try {
      await fsp.access(\`\${assetPath}.${fileExtension}\`, constants.R_OK | constants.W_OK);
      assetPath = \`\${assetPath}.${fileExtension}\`;
      res.setHeader('Content-Encoding', '${contentEncoding}');
    } catch {}
  }

  return fsp.readFile(assetPath);
}`
          },
          nitro.vfs
        ) as never
      )

      nitro.hooks.hook('compiled', (nitro) => {
        const [nodeServer] = globbySync('**/node-server.mjs', {
          cwd: nitro.options.output.dir,
          absolute: true
        })

        if (nodeServer) {
          let nodeServerContent = readFileSync(nodeServer, 'utf8')
          nodeServerContent = nodeServerContent.replace(
            /readAsset\(id\)/g,
            'readAsset(id, event.res)'
          )
          writeFileSync(nodeServer, nodeServerContent)
        }
      })
    })

    nuxt.hook('build:done', () => {})
  }
})
