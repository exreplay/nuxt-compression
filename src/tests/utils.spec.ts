/* eslint-disable no-template-curly-in-string */
import { describe, expect, it } from 'vitest'
import { ModuleOptions } from '../module'
import {
  getCacheHeaders,
  getExtensionAndEncoding,
  getFilterCondition,
  getPublicAssets,
  getPublicAssetsConstants
} from '../utils'

describe('utils', () => {
  it('getPublicAssets should match snapshot', () => {
    const options: ModuleOptions = {
      maxAge: 60 * 60 * 24 * 365,
      viteCompression: {
        algorithm: 'brotliCompress',
        filter: /\.(js|mjs|json|css|html)$/i
      }
    }

    expect(getPublicAssets(options)).toMatchSnapshot()
  })

  it('getExtensionAndEncoding should return correct file extnsion and encoding', () => {
    let { fileExtension, contentEncoding } = getExtensionAndEncoding({
      algorithm: 'gzip'
    })

    expect(fileExtension).toBe('gz')
    expect(contentEncoding).toBe('gzip');

    ({ fileExtension, contentEncoding } = getExtensionAndEncoding({
      algorithm: 'brotliCompress'
    }))

    expect(fileExtension).toBe('br')
    expect(contentEncoding).toBe('br')
  })

  it('getCacheHeaders should return correct setHeader function call', () => {
    let cacheControlHeader = getCacheHeaders({ maxAge: 60 * 60 * 24 * 365 })
    expect(cacheControlHeader).toBe(
      'res.setHeader("Cache-Control", `max-age=${maxAge}, immutable`);'
    )

    cacheControlHeader = getCacheHeaders({
      maxAge: () => 60 * 60 * 24 * 365
    })
    // eslint-disable-next-line no-template-curly-in-string
    expect(cacheControlHeader).toBe(
      'res.setHeader("Cache-Control", `max-age=${maxAge(assetPath)}, immutable`);'
    )

    cacheControlHeader = getCacheHeaders({
      cacheControl: 'public, max-age=60'
    })
    // eslint-disable-next-line no-template-curly-in-string
    expect(cacheControlHeader).toBe(
      'res.setHeader("Cache-Control", cacheControl);'
    )

    cacheControlHeader = getCacheHeaders({
      cacheControl: () => 'public, max-age=60'
    })
    // eslint-disable-next-line no-template-curly-in-string
    expect(cacheControlHeader).toBe(
      'res.setHeader("Cache-Control", cacheControl(assetPath));'
    )

    cacheControlHeader = getCacheHeaders({
      maxAge: 60 * 60 * 24 * 365,
      cacheControl: 'public, max-age=60'
    })
    // eslint-disable-next-line no-template-curly-in-string
    expect(cacheControlHeader).toBe(
      'res.setHeader("Cache-Control", cacheControl);'
    )
  })

  it('getFilterCondition should return correct filter condition', () => {
    let condition = getFilterCondition({
      viteCompression: {
        filter: /\.js$/
      }
    })

    expect(condition).toBe('filter.test(assetPath)')

    condition = getFilterCondition({
      viteCompression: {
        filter: () => true
      }
    })

    expect(condition).toBe('filter(assetPath)')
  })

  it('getPublicAssetsConstants should return correct constants', () => {
    let publicAssetsConstants = getPublicAssetsConstants({
      maxAge: 60 * 60 * 24 * 365,
      viteCompression: {}
    })

    expect(publicAssetsConstants).toBe(
      `
const maxAge = ${60 * 60 * 24 * 365};
const cacheControl = undefined;
const filter = undefined;
      `.trim()
    )

    publicAssetsConstants = getPublicAssetsConstants({
      maxAge: (path) => {
        if (path.includes('js')) {
          return 1
        } else {
          return 2
        }
      },
      viteCompression: {}
    })

    expect(publicAssetsConstants).toBe(
      `
const maxAge = (path) => {
        if (path.includes("js")) {
          return 1;
        } else {
          return 2;
        }
      };
const cacheControl = undefined;
const filter = undefined;
        `.trim()
    )

    publicAssetsConstants = getPublicAssetsConstants({
      cacheControl: 'public, max-age=60',
      viteCompression: {}
    })

    expect(publicAssetsConstants).toBe(
      `
const maxAge = undefined;
const cacheControl = 'public, max-age=60';
const filter = undefined;
      `.trim()
    )

    publicAssetsConstants = getPublicAssetsConstants({
      cacheControl: (path) => {
        if (path.includes('js')) {
          return 'public, max-age=1'
        } else {
          return 'public, max-age=2'
        }
      },
      viteCompression: {}
    })

    expect(publicAssetsConstants).toBe(
      `
const maxAge = undefined;
const cacheControl = (path) => {
        if (path.includes("js")) {
          return "public, max-age=1";
        } else {
          return "public, max-age=2";
        }
      };
const filter = undefined;
        `.trim()
    )

    publicAssetsConstants = getPublicAssetsConstants({
      viteCompression: {
        filter: /\.(js|mjs|json|css|html)$/i
      }
    })

    expect(publicAssetsConstants).toBe(
      `
const maxAge = undefined;
const cacheControl = undefined;
const filter = /\\.(js|mjs|json|css|html)$/i;
        `.trim()
    )

    publicAssetsConstants = getPublicAssetsConstants({
      viteCompression: {
        filter: (path) => {
          if (path.includes('js')) {
            return true
          } else {
            return false
          }
        }
      }
    })

    expect(publicAssetsConstants).toBe(
      `
const maxAge = undefined;
const cacheControl = undefined;
const filter = (path) => {
          if (path.includes("js")) {
            return true;
          } else {
            return false;
          }
        };
        `.trim()
    )
  })
})
