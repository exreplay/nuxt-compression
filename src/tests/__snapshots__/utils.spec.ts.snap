// Vitest Snapshot v1

exports[`utils > getPublicAssets should match snapshot 1`] = `
"
import { promises as fsp, constants } from 'fs'
import { resolve } from 'pathe'
import { dirname } from 'pathe'
import { fileURLToPath } from 'url'
import assets from '#internal/nitro/virtual/public-assets-data'

const maxAge = 31536000;
const cacheControl = undefined;
const filter = /\\\\.(js|mjs|json|css|html)$/i;
  
export async function readAsset (id, res) {
  const serverDir = dirname(fileURLToPath(import.meta.url))
  let assetPath = resolve(serverDir, assets[id].path)

  if (filter.test(assetPath)) {
    try {
      await fsp.access(\`\${assetPath}.br\`, constants.R_OK | constants.W_OK);
      assetPath = \`\${assetPath}.br\`;
      res.setHeader('Content-Encoding', 'br');
    } catch {}
  }

  res.setHeader(\\"Cache-Control\\", \`max-age=\${maxAge}, immutable\`);

  return fsp.readFile(assetPath);
}"
`;
